docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,264 @@
+
+### Kata-agent changes
+
+For mounting a filesystem on the block device volume, the information can be passed to Kata-agent through the existing 
+`Storage` grpc object. Two new grpc request objects are added to the Kata-agent grpc interface for 
+resizing and getting volume stats:
+```Golang
+type ResizeFileSystemRequest struct {
+  mountPoint string
+  size       int64
+}
+
+type GetVolumeStatsRequest struct {
+  mountPoint string
+}
+```
+The Kata agent then handles the requests and returns the information back to the kata shim.
+
+### Step by step walkthrough
+
+Given the following definition:
+```YAML
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: app
+spec:
+  runtime-class: kata-qemu
+  containers:
+  - name: app
+    image: centos
+    command: ["/bin/sh"]
+    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
+    volumeMounts:
+    - name: persistent-storage
+      mountPath: /data
+  volumes:
+  - name: persistent-storage
+    persistentVolumeClaim:
+      claimName: ebs-claim
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: ebs-claim
+spec:
+  accessModes:
+    - ReadWriteOncePod
+  volumeMode: Filesystem
+  storageClassName: ebs-sc
+  resources:
+    requests:
+      storage: 4Gi
+---
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: ebs-sc
+provisioner: ebs.csi.aws.com
+volumeBindingMode: WaitForFirstConsumer
+parameters:
+  csi.storage.k8s.io/fstype: ext4
+
+```
+Let’s assume that changes have been made in aws-ebs-csi-driver node driver.
+
+**Node publish volume**
+
+1. In the node csi driver, the NodePublishVolume API invokes: `kata-runtime fs add-blk-mount-info “{device: /dev/sdf,fstype: ext4}”`.
+2. Kata-runtime writes the json to a file called `mount-info.json` with the above json under `/run/kata-containers/shared/direct-block-devices/sdf/`.
+
+**Node unstage volume**
+
+1. In the node csi driver, the NodeUnstageVolume API invokes: `kata-runtime fs reset-blk-mount-info "/dev/sdf"`.
+2. Kata-runtime deletes the directory `/run/kata-containers/shared/direct-block-devices/sdf/`.
+
+**Use the volume in sandbox**
+1. Upon the request to start a sandbox, the containerd-shim-kata-v2(shim for abbreviation) examines the container spec 
+and iterates through the mounts, for each mount, if the device name maps to a directory under `/run/kata-containers/shared/direct-block-devices`, 
+it then generates a `storage` grpc object combining the mount info file and the mount attributes in the container spec.
+2. The shim sends the storage objects to kata-agent through ttrpc.
+3. The shim writes a file with the sandbox id as the name under `/run/kata-containers/shared/direct-block-devices/[device_name]`.
+4. The kata-agent mounts the storage objects into the containers.
+
+**Node expand volume**
+1. In the node csi driver, the NodeExpandVolume API invokes: `kata-runtime fs resize-volume –-device-path /dev/sdf –-size 8Gi`
```

Comment:
All other commands are using `/dev/sdf` as volume identifier. The `--device-path` parameter does not seem necessary.